Ktisis/Interface/Windows/ActorEdit/EditEquip.cs

```diff
@@ -125,30 +144,152 @@ public class EditEquip {
 				var focus = ImGui.IsWindowFocused() || ImGui.IsWindowHovered();
 
 				ImGui.PushItemWidth(400);
-				ImGui.InputTextWithHint("##equip_search", "Search...", ref ItemSearch, 32);
+				bool pressedEnter = ImGui.InputTextWithHint("##equip_search", "Search...", ref ItemSearch, 32, ImGuiInputTextFlags.EnterReturnsTrue);
+				if (ImGui.IsWindowFocused(ImGuiFocusedFlags.RootAndChildWindows) && !ImGui.IsAnyItemActive() && !ImGui.IsMouseClicked(ImGuiMouseButton.Left))
+					ImGui.SetKeyboardFocusHere(-1);
+
 				ImGui.BeginListBox("##equip_items", new Vector2(-1, 300));
+
 				var items = SlotItems;
 				if (ItemSearch.Length > 0)
-					items = items.Where(i => i.Name.Contains(ItemSearch));
+					items = items.Where(i => i.Name.Contains(ItemSearch, StringComparison.OrdinalIgnoreCase));
+
+				int indexKey = 0;
+				bool isOneSelected = false; // allows one selection per foreach
+				if (LastSelectedItemKey >= items.Count()) LastSelectedItemKey = items.Count() - 1;
+
 				foreach (var item in items) {
 					// TODO: Icon?
-					if (ImGui.Selectable($"{item.Name}")) {
+
+					bool selecting = false;
+
+					selecting |= ImGui.Selectable($"{item.Name}", indexKey == LastSelectedItemKey);
+					if (!isOneSelected)
+					{
+						selecting |= ImGui.IsKeyPressed(KeyBindBrowseUp) && indexKey == LastSelectedItemKey - 1;
+						selecting |= ImGui.IsKeyPressed(KeyBindBrowseDown) && indexKey == LastSelectedItemKey + 1;
+						selecting |= ImGui.IsKeyPressed(KeyBindBrowseUpFast) && indexKey == LastSelectedItemKey - FastScrollLineJump;
+						selecting |= ImGui.IsKeyPressed(KeyBindBrowseDownFast) && indexKey == LastSelectedItemKey + FastScrollLineJump;
+						selecting |= pressedEnter;
+					}
+
+					if (selecting) {
+						if (ImGui.IsKeyPressed(KeyBindBrowseUp) || ImGui.IsKeyPressed(KeyBindBrowseDown) || ImGui.IsKeyPressed(KeyBindBrowseUpFast) || ImGui.IsKeyPressed(KeyBindBrowseDownFast))
+							ImGui.SetScrollY(ImGui.GetCursorPosY() - (ImGui.GetWindowHeight() /2));
 						equip.Id = item.Model.Id;
 						equip.Variant = (byte)item.Model.Variant;
 						Target->Equip(index, equip);
+						LastSelectedItemKey = indexKey;
+						isOneSelected = true;
 					}
 					focus |= ImGui.IsItemFocused();
+					indexKey++;
 				}
 				ImGui.EndListBox();
 				focus |= ImGui.IsItemActive();
 				ImGui.PopItemWidth();
 
-				if (!focus) {
+				if (!focus || ImGui.IsKeyPressed(ImGuiKey.Escape)) {
 					SlotSelect = null;
 					SlotItems = null;
 				}
 			}
 
+			ImGui.End();
+		}
+		public static void DrawControls()
+		{
+			Sets = EquipmentSets.Init(Items);
+
+			if (GuiHelpers.IconButtonTooltip(FontAwesomeIcon.Tshirt, "Look up for a set."))
+				OpenSetSelector();
+
+			if (DrawSetSelection)
+				DrawSetSelectorList();
+
+			ImGui.Separator();
+		}
+
+
+
+
+		// dirty shameless copy paste of DrawSelectorList(), merge common code?
```

Comment:
^

Reply:
Well for this, the two methods have become different, I tried making only one with dynamic parameters but I couldn't cast them or use use .Where() once inside the method. After some struggling I gave up and reverted the changes.

Also it ended up having one callback to draw the line and another one when a line is selected selected.

Reply:
Might look at factoring out the common code this evening, though for... reasons you'll have to wait until tomorrow for any results. (Feel free to push changes in the meanwhile, though)

Otherwise... drop the comment.

Reply:
I'm going to try the following.
I thought of making methods like `BeginPopupHooverWindow` and `EndPopupHooverWindow` with returning the `ImGui.Begin` bool on begin, and pass a few variables (like `focus`) as `ref`.
And maybe make flags to make them either `ImGui` `ListBox`, or other kind of window.